### PR TITLE
May now be merged

### LINKED ISSUE #11110: Add LaForce THM benchmark to PorousFlow
## Rationale

We have benchmarked PorousFlow vs a complicated analytical solution to a nontrivial dynamic THM problem derived by LaForce et al.  Add the test and doco to PorousFlow

## Description

This is a "tour de la force", heheheh, of PorousFlow's THM capability.  PorousFlow agrees extremely well with the analytical solution and it'll be very instructive for users to see the agreement and how the input file is constructed.  We are still awaiting the "OK" from our funders before we can add to the repo, however.  @cpgr, do you want to prompt JEK about this?

## Impact

Perfect example of PorousFlow for the repo - not a complicated mesh, but super complicated physics and analytical solution.

### PR DESCRIPTION
**Pull request by WilkAndy**
Cold injection of CO2 benchmark THM problem added to PorousFlow

Fixes #11110

### CODE REVIEW

**Comment by WilkAndy** [comment]
Please do not merge yet.  We are seeking permission from a funding agency.

**Comment by WilkAndy** [comment]
Perhaps @cpgr would like to review?

**Comment by cpgr** [comment]
I'll read through it tonight

**Comment by cpgr** [comment]
Hey @WilkAndy - I didn't get to it last night as I had no power at home due to the wild weather. I'll take a look today after it finishes building

**Comment by WilkAndy** [comment]
Woohoo, wild weather sounds frightening.  I don't think there's a great rush for this PR, because JEK has to ask ANLEC anyway (remind him!).  But the review shouldn't take you too long anyway - you've seen it all before.

File: modules/porous_flow/doc/content/modules/porous_flow/thm_example.md [hunk 1]
```diff
@@ -0,0 +1,519 @@
+# Cold CO$_{2}$ injection into an elastic reservoir - a multi-phase THM problem
+
+Andy Wilkins, Chris Green, Jonathon Ennis-King
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/thm_example.md]: Jonathan with an a!

File: modules/porous_flow/doc/content/modules/porous_flow/thm_example.md [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+\begin{aligned}
+  \rho C & = (1 - \phi)\rho_{R}C_{R} + \phi
+  S_{\mathrm{water}}\rho_{\mathrm{water}}C_{\mathrm{water}} + \phi (1
+  - S_{\mathrm{water}}) \rho_{\mathrm{co2}}C_{\mathrm{co2}} \ , \\
+  \rho C \mathbf{Q} & = \rho_{\mathrm{water}} C_{\mathrm{water}}
+  {\mathbf{Q}}_{\mathrm{water}}  + \rho_{\mathrm{co2}} C_{\mathrm{co2}}
+  {\mathbf{Q}}_{\mathrm{co2}} \ .
+\end{aligned}
+\end{equation}
+[simplified.heat.eqn] and [defns.rhoc.rhocq.eqn] are exactly Equation~(A1) in [!citet](laforce2014a).
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/thm_example.md]: Maybe change to Eq. (A1) for consistency?

File: modules/porous_flow/doc/content/modules/porous_flow/thm_example.md [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+
+Then a [PressureBC](bcs/Pressure.md) applies this *total* stress (physically, a mechanical pushing force) at the borehole wall:
+
+!listing modules/porous_flow/examples/thm_example/2D.i start=[./cavity_pressure_x] end=[]
+
+The `component` is zero because this is an axially-symmetric (2D RZ) problem.  The value of the total pressure should be equal to the variable `pwater`, but because the `p_bh` is recorded at the start of each time step, this boundary condition actually lags by one time step.  Because porepressure `pwater` does not change significantly over each time step, this leads to minimal error.
+
+## Results
+
+!media media/porous_flow/2D_thm_compare_porepressure_fig.png caption=Comparison between the PorousFlow result and the analytic expression derived by La Force et al. for the porepressure id=porepressure_cf_fig
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/thm_example.md]: Maybe add the Eq. number link in the caption (same for other figures) to make it easier for people to get back to the equation?

File: modules/porous_flow/examples/thm_example/paper_solns.py [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+    xd = v_saturation_shock * td
+    return r_given_xd(xd)
+
+def temperature_shock_position(t, mu1, mu2):
+    td = td_given_t(t)
+    v_temperature_shock = df1_dss1(ss1tt(mu1, mu2), mu1, mu2)
+    xd = v_temperature_shock * td
+    return r_given_xd(xd)
+
+#print "temp front: one hour", temperature_shock_position(one_hour, mu1_moose, mu2_moose)
```
Comment (cpgr) [inline modules/porous_flow/examples/thm_example/paper_solns.py]: Do you want these commented print statements?

File: modules/porous_flow/doc/content/modules/porous_flow/thm_example.md [hunk 5]
```diff
@@ -0,0 +1,519 @@
+# Cold CO$_{2}$ injection into an elastic reservoir - a multi-phase THM problem
+
+Andy Wilkins, Chris Green, Jonathon Ennis-King
+
+Two papers of LaForce et al. [!citep](laforce2014a,laforce2014b) derive semi-analytical solutions of a 2-phase THM problem involving injection of a cold fluid into a warmer, elastic reservoir depicted in [tara_setup_fig].
+In this page, their solution will be replicated in MOOSE.  The nomenclature used here is described in a [separate page](nomenclature.md).
+
+!media media/porous_flow/tara_setup.png style=width:30% caption=The geometrical setup used in the papers of LaForce et al.  id=tara_setup_fig
+
+LaForce et al. consider a cylindrically-symmetric model, as depicted in [tara_setup_fig] , with radial coordinate labelled by $r$ and axial coordinate $z$.  A permeable reservoir is sandwiched between two impermeable seals.  The reservoir and seals are oriented horizontally (normal to $z$).  The reservoir is initally fully water saturated.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/thm_example.md]: Remove space before comma

Comment (cpgr) [reply modules/porous_flow/doc/content/modules/porous_flow/thm_example.md]: initally -> initially

**Review by cpgr — COMMENTED** [review]
Looks great Andy. I only had minor comments, but I think that this example will really help people set up complicated problems as it explains things really well.

**Comment by WilkAndy** [comment]
Thanks @cpgr .   My eyesight is really bad.  I don't know how long i've been spelling Jonathan's name incorrectly!!   Oh dear, he never told me.

**Comment by WilkAndy** [comment]
This may now be merged as we have gained permission from a funding agency

**Review by permcody — APPROVED** [review]

**Pull request merged**